File: 백준/Gold/13265. 색칠하기/색칠하기.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] visited;
    static ArrayList<Integer>[] graph;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int T = Integer.parseInt(br.readLine());

        for (int i = 0; i < T; i++) {
            st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int M = Integer.parseInt(st.nextToken());
            visited = new int[N + 1];

            graph = new ArrayList[N + 1];
            for (int j = 0; j < N + 1; j++) {
                graph[j] = new ArrayList<>();
            }

            for (int j = 0 ; j < M; j++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                graph[x].add(y);
                graph[y].add(x);
            }

            boolean check = false;
            for (int j = 1; j < N+1; j++) {
                if (visited[j] > 0) continue;

                if (!dfs(j)) {
                    System.out.println("impossible");
                    check = true;
                    break;
                }
            }
            if (!check) System.out.println("possible");
        }
    }

    static boolean dfs(int i) {
        ArrayDeque<int[]> stack = new ArrayDeque<>();
        stack.push(new int[] {i,1});

        while (!stack.isEmpty()) {
            int[] tmp = stack.pop();
            int cur = tmp[0];
            int cnt = tmp[1];

            if (visited[cur] > 0) {
                continue;
            } else {
                visited[cur] = cnt;
            }

            for (int next : graph[cur]) {
                if (visited[next] > 0 && (visited[next] - cnt) % 2 == 0) {
                    return false;
                }
                stack.push(new int[] {next,cnt+1});
            }
        }

        return true;
    }
}

File: 백준/Gold/13265. 색칠하기/색칠하기.py
# 색칠하기
import sys
input = sys.stdin.readline

def dfs(n):
    stack = [(n, 1)]

    while stack:
        cur, cnt = stack.pop()

        if visited[cur] != 0:
            continue
        else:
            visited[cur] = cnt

        for j in graph[cur]:
            if visited[j] == cnt:
                return False

            if visited[j] != 0:
                continue
            else:
                stack.append((j, -cnt))

    return True

T = int(input())

for i in range(T):
    N, M = map(int,input().split())
    graph = [[] for _ in range(N+1)]
    visited = [0] * (N+1)

    for j in range(M):
        s, t = map(int,input().split())
        graph[s].append(t)
        graph[t].append(s)

    check = True
    for i in range(1, N+1):
        if not dfs(i):
            check = False
            break

    if not check:
        print("impossible")
    else:
        print("possible")
    
